### PR TITLE
Logic/commands: implement `planner_suggest` command

### LINKED ISSUE #191: As a user, I want the degree plan to suggest to me what modules to take

### PR DESCRIPTION
**Pull request by WendyBaiYunwei**
Closes #191.

Currently, users may not know what module(s) to put in the degree plan. 
Let's implement `planner_suggest` to facilitate recommendations of module(s) to add to the degree plan.
* `planner_suggest` requires users to input desirable credits and tags, and module(s) with the highest number of matching tags and does not exist in the degree plan will be suggested first. 
* If tie, the module(s) with closest credits will be suggested first. 
* Tag-matching is prioritized in the recommendation; if there is no tag matched, no module will be shown.
* `tag` is optional parameter. If no `tag` is in input, the recommendation will be based on credits.
* modules with matching tags and credits that are not in the degree plan are given as supplementary information.
* user guide will be updated through another PR.

### CODE REVIEW

**Comment by WendyBaiYunwei** [comment]
Planner_suggest has compulsory field "tag". This design is bad because not every module has a tag.
Let's make "tag" optional.

File: src/main/java/pwe/planner/logic/commands/PlannerSuggestCommand.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    @Override
+    public CommandResult execute(Model model, CommandHistory history) throws CommandException {
+        requireNonNull(model);
+
+        Set<Code> plannerCodes = new HashSet<>();
+        model.getApplication().getDegreePlannerList()
+                .stream().map(DegreePlanner::getCodes).forEach(plannerCodes::addAll);
+
+        ObservableList<Module> moduleList = model.getApplication().getModuleList();
+        List<ModuleToSuggest> modulesToSuggest = new ArrayList<ModuleToSuggest>();
```
Comment (lycjackie) [inline src/main/java/pwe/planner/logic/commands/PlannerSuggestCommand.java]: I believe the RHS of `<ModuleToSuggest>` is not needed here. (Mentioned by Intellij)

Same for the next few declaration

File: src/main/java/pwe/planner/logic/commands/PlannerSuggestCommand.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        codesWithMatchingTags.removeAll(plannerCodes);
+
+        //Returns codes with matching credits.
+        List<Code> codesWithMatchingCredits = new ArrayList<>();
+        for (ModuleToSuggest moduleWithMatchingCredits : modulesWithMatchingCredits) {
+            codesWithMatchingCredits.add(moduleWithMatchingCredits.getModuleCode());
+        }
+        codesWithMatchingCredits.removeAll(plannerCodes);
+
+        model.commitApplication();
```
Comment (lycjackie) [inline src/main/java/pwe/planner/logic/commands/PlannerSuggestCommand.java]: Why would you need to do this when there is no modification? 

Correct me if I'm wrong but planner_suggest is not doing any mutation of the data?

File: src/main/java/pwe/planner/logic/commands/PlannerSuggestCommand.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            if (!matchingTags.isEmpty()) {
+                modulesWithMatchingTags.add(moduleToSuggest);
+            }
+
+            if (creditDifference == 0) {
+                modulesWithMatchingCredits.add(moduleToSuggest);
+            }
+        }
+
+        modulesToSuggest.sort(new SortModulesToSuggest());
```
Comment (lycjackie) [inline src/main/java/pwe/planner/logic/commands/PlannerSuggestCommand.java]: Do you really need a comparator class just to do this?

You could just create a private method to do the comparator/lambda function or you could just implement comparator directly inside the `ModuleToSuggest` class. 

So you could do `modulesToSuggest.sort()` directly.

File: src/main/java/pwe/planner/logic/commands/PlannerSuggestCommand.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            codesToSuggest.add(moduleToSuggest.getModuleCode());
+        }
+        codesToSuggest.removeAll(plannerCodes);
+        List<Code> truncatedList = codesToSuggest.subList(0, min(codesToSuggest.size(), MAX_NUMBER_OF_ELEMENETS));
+
+        //Returns codes with matching tags.
+        List<Code> codesWithMatchingTags = new ArrayList<>();
+        for (ModuleToSuggest moduleWithMatchingTags : modulesWithMatchingTags) {
+            codesWithMatchingTags.add(moduleWithMatchingTags.getModuleCode());
+        }
```
Comment (lycjackie) [inline src/main/java/pwe/planner/logic/commands/PlannerSuggestCommand.java]: The whole method actually can be simplified using the following code.
```
List<Code> codesWithMatchingTags = modulesWithMatchingTags.stream()
                .map(ModuleToSuggest::getModuleCode).collect(Collectors.toList());
```

File: src/main/java/pwe/planner/logic/commands/PlannerSuggestCommand.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    public PlannerSuggestCommand(Credits bestCredits, Set<Tag> tags) {
+        requireAllNonNull(bestCredits, tags);
+
+        creditsToFind = bestCredits;
+        tagsToFind = tags;
+    }
+
+    @Override
+    public CommandResult execute(Model model, CommandHistory history) throws CommandException {
```
Comment (lycjackie) [inline src/main/java/pwe/planner/logic/commands/PlannerSuggestCommand.java]: `execute(...)` throws `CommandExceptions` yet is never throw? :O

**Review by lycjackie — CHANGES_REQUESTED** [review]

**Comment by lycjackie** [comment]
When testing out the command, the Module(s) with matching credits is not limiting to `MAX_NUMBER_OF_ELLEMENTS`

image

Is this intended?

File: src/main/java/pwe/planner/model/planner/ModuleToSuggest.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package pwe.planner.model.planner;
+
+import pwe.planner.model.module.Code;
+
+/**
+ * Creates a ModuleToSuggest object for sorting.
+ */
+public class ModuleToSuggest {
```
Comment (lycjackie) [inline src/main/java/pwe/planner/model/planner/ModuleToSuggest.java]: Actually, for this class, do you think is possible to just use `Credits`, `Tags` and `Code` class directly?

Alternatively, maybe you want to consider an inner class to help wrap around the attributes you need instead? 

As far as I can see, there are no other codes that are using this model.

**Review by lycjackie — CHANGES_REQUESTED** [review]

**Comment by WendyBaiYunwei** [comment]
> When testing out the command, the Module(s) with matching credits is not limiting to `MAX_NUMBER_OF_ELLEMENTS`
> 
> image
> 
> Is this intended?

The list of recommended modules is intended to be limited to 10 elements, but the modules with matching credits and tags are not limit to 10. Do you think it is acceptable?

**Comment by lycjackie** [comment]
> The list of recommended modules is intended to be limited to 10 elements, but the modules with matching credits and tags are not limit to 10. Do you think it is acceptable?

Since the intention was to suggest to the user, maybe a filtered list would be better?

**Review by lycjackie — APPROVED** [review]

**Review by Letm3through — APPROVED** [review]

**Pull request merged**